File: ios/Runner/FlutterYodo1Mas.m
//
//  FlutterYodo1Mas.m
//  Runner
//
//  Created by 周玉震 on 2021/9/17.
//

#import "FlutterYodo1Mas.h"
#import <Yodo1MasCore/Yodo1Mas.h>
#import <YYModel/YYModel.h>

#define CHANNEL @"com.yodo1.mas/sdk"
#define METHOD_NATIVE_INIT_SDK @"native_init_sdk"
#define METHOD_NATIVE_IS_AD_LOADED @"native_is_ad_loaded"
#define METHOD_NATIVE_SHOW_AD @"native_show_ad"

#define METHOD_FLUTTER_INIT_EVENT @"flutter_init_event"
#define METHOD_FLUTTER_AD_EVENT @"flutter_ad_event"

@interface FlutterYodo1Mas()<Yodo1MasRewardAdDelegate, Yodo1MasInterstitialAdDelegate, Yodo1MasBannerAdDelegate>

@property (nonatomic, strong) UIViewController *controller;
@property (nonatomic, strong)  FlutterMethodChannel *channel;

@end


@implementation FlutterYodo1Mas {
    
}

+ (FlutterYodo1Mas *)sharedInstance {
    static FlutterYodo1Mas *_instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [[FlutterYodo1Mas alloc] init];
    });
    return _instance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        [Yodo1Mas sharedInstance].rewardAdDelegate = self;
    }
    return self;
}

- (void)buildWithController:(FlutterViewController *)controller {
    __weak __typeof(self)weakSelf = self;
    _controller = controller;
    _channel = [FlutterMethodChannel methodChannelWithName:CHANNEL binaryMessenger:controller];
    [_channel setMethodCallHandler:^(FlutterMethodCall* call, FlutterResult result) {
        if (call.method) {
            if ([call.method isEqualToString: METHOD_NATIVE_INIT_SDK]) {
                NSString *appKey = call.arguments[@"app_key"];
                
                if (appKey) {
                    [weakSelf initSdk: appKey];
                    result(@1);
                }
            } else if ([call.method isEqualToString: METHOD_NATIVE_IS_AD_LOADED]) {
                BOOL isAdLoaded = NO;
                NSString *type = call.arguments[@"ad_type"];
                if (type) {
                    if ([type isEqualToString:@"Reward"]) {
                        isAdLoaded = [[Yodo1Mas sharedInstance] isRewardAdLoaded];
                    } else if ([type isEqualToString:@"Interstitial"]) {
                        isAdLoaded = [[Yodo1Mas sharedInstance] isInterstitialAdLoaded];
                    } if ([type isEqualToString:@"Banner"]) {
                        isAdLoaded = [[Yodo1Mas sharedInstance] isBannerAdLoaded];
                    }
                }
                result(@(isAdLoaded));
            } else if ([call.method isEqualToString: METHOD_NATIVE_SHOW_AD]) {
                NSLog(@"Show Ad - %@", call.arguments);
                NSString *type = call.arguments[@"ad_type"];
                if (type) {
                    if ([type isEqualToString:@"Reward"]) {
                        [[Yodo1Mas sharedInstance] showRewardAd];
                    } else if ([type isEqualToString:@"Interstitial"]) {
                        [[Yodo1Mas sharedInstance] showInterstitialAd];
                    } if ([type isEqualToString:@"Banner"]) {
                        [[Yodo1Mas sharedInstance] showBannerAd];
                    }
                }
                result(nil);
            } else {
                result(FlutterMethodNotImplemented);
            }
        } else {
            result(FlutterMethodNotImplemented);
        }
    }];
}

- (void)initSdk:(NSString *)appKey {
    __weak __typeof(self)weakSelf = self;
    [[Yodo1Mas sharedInstance] initWithAppKey:appKey successful:^{
        if (weakSelf.channel) {
            NSDictionary *json = @{@"successful" : @(YES)};
            [weakSelf.channel invokeMethod:METHOD_FLUTTER_INIT_EVENT arguments: json];
        }
    } fail:^(Yodo1MasError * error) {
        if (weakSelf.channel) {
            NSDictionary *json = @{@"successful" : @(NO), @"error" : [error getJsonObject]};
            [weakSelf.channel invokeMethod:METHOD_FLUTTER_INIT_EVENT arguments: json];
        }
    }];
}

#pragma mark - Yodo1Masdelegate
- (void)onAdOpened:(Yodo1MasAdEvent *)event {
    if (self.channel) {
        [self.channel invokeMethod:METHOD_FLUTTER_AD_EVENT arguments: [event getJsonObject]];
    }
}

- (void)onAdClosed:(Yodo1MasAdEvent *)event {
    if (self.channel) {
        [self.channel invokeMethod:METHOD_FLUTTER_AD_EVENT arguments: [event getJsonObject]];
    }
}

- (void)onAdError:(Yodo1MasAdEvent *)event error:(Yodo1MasError *)error {
    if (self.channel) {
        [self.channel invokeMethod:METHOD_FLUTTER_AD_EVENT arguments: [event getJsonObject]];
    }
}

- (void)onAdRewardEarned:(Yodo1MasAdEvent *)event {
    if (self.channel) {
        [self.channel invokeMethod:METHOD_FLUTTER_AD_EVENT arguments: [event getJsonObject]];
    }
}

@end

File: Integration/Java/FlutterYodo1Mas.java
import android.app.Activity;
import android.util.Log;

import androidx.annotation.NonNull;

import com.google.gson.JsonObject;
import com.yodo1.mas.Yodo1Mas;
import com.yodo1.mas.error.Yodo1MasError;
import com.yodo1.mas.event.Yodo1MasAdEvent;
import com.yodo1.mas.helper.model.Yodo1MasAdBuildConfig;

import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.FlutterEngine;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.Result;

public class FlutterYodo1Mas {

   private static FlutterYodo1Mas helper;
   public static FlutterYodo1Mas getInstance() {
       if (helper == null) {
           synchronized (Yodo1Mas.class) {
               if (helper == null) {
                   helper = new FlutterYodo1Mas();
               }
           }
       }
       return helper;
   }

   private static final String CHANNEL = "com.yodo1.mas/sdk";
   private static final String METHOD_NATIVE_INIT_SDK = "native_init_sdk";
   private static final String METHOD_NATIVE_IS_AD_LOADED = "native_is_ad_loaded";
   private static final String METHOD_NATIVE_SHOW_AD = "native_show_ad";

   private static final String METHOD_FLUTTER_INIT_EVENT = "flutter_init_event";
   private static final String METHOD_FLUTTER_AD_EVENT = "flutter_ad_event";
   private static final String DISMISS_BANNER = "dismiss_banner";

   private MethodChannel channel;
   private Activity activity;

   private FlutterYodo1Mas() {
       Yodo1Mas.getInstance().setRewardListener(new Yodo1Mas.RewardListener() {
           @Override
           public void onAdOpened(@NonNull Yodo1MasAdEvent event) {
               if (channel != null) {
                   channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, event.getJSONObject().toString());
               }
           }

           @Override
           public void onAdError(@NonNull Yodo1MasAdEvent event, @NonNull Yodo1MasError error) {
               if (channel != null) {
                   channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, event.getJSONObject().toString());
               }
           }

           @Override
           public void onAdClosed(@NonNull Yodo1MasAdEvent event) {
               if (channel != null) {
                   channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, event.getJSONObject().toString());
               }
           }

           @Override
           public void onAdvertRewardEarned(@NonNull Yodo1MasAdEvent event) {
               if (channel != null) {
                   channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, event.getJSONObject().toString());
               }
           }
       });

       Yodo1Mas.getInstance().setInterstitialListener(new Yodo1Mas.InterstitialListener() {
           @Override
           public void onAdOpened(@NonNull Yodo1MasAdEvent event) {
               if (channel != null) {
                   channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, event.getJSONObject().toString());
               }
           }

           @Override
           public void onAdError(@NonNull Yodo1MasAdEvent event, @NonNull Yodo1MasError error) {
               if (channel != null) {
                   channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, event.getJSONObject().toString());
               }
           }

           @Override
           public void onAdClosed(@NonNull Yodo1MasAdEvent event) {
               if (channel != null) {
                   channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, event.getJSONObject().toString());
               }
           }
       });
       Yodo1Mas.getInstance().setBannerListener(new Yodo1Mas.BannerListener() {
           @Override
           public void onAdOpened(@NonNull Yodo1MasAdEvent event) {
               if (channel != null) {
                   channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, event.getJSONObject().toString());
               }
           }

           @Override
           public void onAdError(@NonNull Yodo1MasAdEvent event, @NonNull Yodo1MasError error) {
               if (channel != null) {
                   channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, event.getJSONObject().toString());
               }
           }

           @Override
           public void onAdClosed(@NonNull Yodo1MasAdEvent event) {
               if (channel != null) {
                   channel.invokeMethod(METHOD_FLUTTER_AD_EVENT, event.getJSONObject().toString());
               }
           }
       });
   }

   public void build(FlutterEngine flutterEngine, Activity activity) {
       this.activity = activity;
       this.channel = new MethodChannel(flutterEngine.getDartExecutor().getBinaryMessenger(), CHANNEL);
       this.channel.setMethodCallHandler(
               new MethodChannel.MethodCallHandler() {
                   @Override
                   public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
                       switch (call.method) {
                           case METHOD_NATIVE_INIT_SDK: {
                               String appKey = call.argument("app_key");
                               if (appKey != null) {
                                   initSdk(appKey);
                               }
                               result.success(null);
                               break;
                           }
                           case METHOD_NATIVE_IS_AD_LOADED:{
                               boolean isAdLoaded = false;
                               String type = call.argument("ad_type");
                               if (type != null) {
                                   switch (type) {
                                       case "Reward": {
                                           isAdLoaded = Yodo1Mas.getInstance().isRewardedAdLoaded();
                                           break;
                                       }
                                       case "Interstitial": {
                                           isAdLoaded = Yodo1Mas.getInstance().isInterstitialAdLoaded();
                                           break;
                                       }
                                       case "Banner": {
                                           isAdLoaded = Yodo1Mas.getInstance().isBannerAdLoaded();
                                           break;
                                       }
                                   }
                               }
                               result.success(isAdLoaded);
                               break;
                           }
                           case METHOD_NATIVE_SHOW_AD: {
                               String type = call.argument("ad_type");
                               if (type != null) {
                                   switch (type) {
                                       case "Reward": {
                                           Yodo1Mas.getInstance().showRewardedAd(FlutterYodo1Mas.this.activity);
                                           break;
                                       }
                                       case "Interstitial": {
                                           Yodo1Mas.getInstance().showInterstitialAd(FlutterYodo1Mas.this.activity);
                                           break;
                                       }
                                       case "Banner": {
                                           Yodo1Mas.getInstance().showBannerAd(FlutterYodo1Mas.this.activity);
                                           break;
                                       }
                                   }
                               }
                               result.success(null);
                               break;
                           }
                           case DISMISS_BANNER: {
                               Yodo1Mas.getInstance().dismissBannerAd();
                               result.success(null);
                               break;
                           }
                       }
                   }
               }
       );
   }

   private void initSdk(@NonNull String appKey) {
       Yodo1MasAdBuildConfig config = new Yodo1MasAdBuildConfig.Builder().enableUserPrivacyDialog(true).build();
       Yodo1Mas.getInstance().setAdBuildConfig(config);
       Yodo1Mas.getInstance().init(activity, appKey, new Yodo1Mas.InitListener() {
           @Override
           public void onMasInitSuccessful() {
               if (channel != null) {
                   JsonObject json = new JsonObject();
                   json.addProperty("successful", true);
                   Log.d(FlutterYodo1Mas.CHANNEL, "调用Flutter方法 - " + METHOD_FLUTTER_INIT_EVENT);
                   Map<String, String> args = new HashMap<>();
                   args.put("successful", "true");
                   channel.invokeMethod(METHOD_FLUTTER_INIT_EVENT, args);
               }
           }

           @Override
           public void onMasInitFailed(@NonNull Yodo1MasError error) {
               if (channel != null) {
                   JsonObject json = new JsonObject();
                   json.addProperty("successful", false);
                   json.addProperty("error", error.getJsonObject().toString());
                   channel.invokeMethod(METHOD_FLUTTER_INIT_EVENT, json.toString());
               }
           }
       });
   }
}